File: src/backend/estoque/src/model/requisicao_interna_cabecalho_model.py
from src import db
from src.model.requisicao_interna_detalhe_model import RequisicaoInternaDetalheModel
from src.model.view_pessoa_colaborador_model import ViewPessoaColaboradorModel


class RequisicaoInternaCabecalhoModel(db.Model):
    __tablename__ = 'requisicao_interna_cabecalho'

    id = db.Column(db.Integer, primary_key=True)
    data_requisicao = db.Column(db.DateTime)
    situacao = db.Column(db.String(1))
    id_colaborador = db.Column(db.Integer, db.ForeignKey('view_pessoa_colaborador.id'))

    requisicao_interna_detalhe_model_list = db.relationship('RequisicaoInternaDetalheModel', lazy='dynamic')
    view_pessoa_colaborador_model = db.relationship('ViewPessoaColaboradorModel', foreign_keys=[id_colaborador])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_colaborador = data.get('idColaborador')
        self.data_requisicao = data.get('dataRequisicao')
        self.situacao = data.get('situacao')

    def serialize(self):
        return {
            'id': self.id,
            'idColaborador': self.id_colaborador,
            'dataRequisicao': self.data_requisicao.isoformat(),
            'situacao': self.situacao,
            'requisicaoInternaDetalheModelList': [requisicao_interna_detalhe_model.serialize() for requisicao_interna_detalhe_model in self.requisicao_interna_detalhe_model_list],
            'viewPessoaColaboradorModel': self.view_pessoa_colaborador_model.serialize() if self.view_pessoa_colaborador_model else None,
        }

File: src/frontend/estoque/lib/app/data/model/requisicao_interna_cabecalho_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:estoque/app/infra/infra_imports.dart';
import 'package:estoque/app/data/model/model_imports.dart';
import 'package:intl/intl.dart';
import 'package:estoque/app/data/domain/domain_imports.dart';

class RequisicaoInternaCabecalhoModel {
	int? id;
	int? idColaborador;
	DateTime? dataRequisicao;
	String? situacao;
	List<RequisicaoInternaDetalheModel>? requisicaoInternaDetalheModelList;
	ViewPessoaColaboradorModel? viewPessoaColaboradorModel;

	RequisicaoInternaCabecalhoModel({
		this.id,
		this.idColaborador,
		this.dataRequisicao,
		this.situacao,
		this.requisicaoInternaDetalheModelList,
		this.viewPessoaColaboradorModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'data_requisicao',
		'situacao',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Data Requisicao',
		'Situacao',
	];

	RequisicaoInternaCabecalhoModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idColaborador = jsonData['idColaborador'];
		dataRequisicao = jsonData['dataRequisicao'] != null ? DateTime.tryParse(jsonData['dataRequisicao']) : null;
		situacao = RequisicaoInternaCabecalhoDomain.getSituacao(jsonData['situacao']);
		requisicaoInternaDetalheModelList = (jsonData['requisicaoInternaDetalheModelList'] as Iterable?)?.map((m) => RequisicaoInternaDetalheModel.fromJson(m)).toList() ?? [];
		viewPessoaColaboradorModel = jsonData['viewPessoaColaboradorModel'] == null ? ViewPessoaColaboradorModel() : ViewPessoaColaboradorModel.fromJson(jsonData['viewPessoaColaboradorModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idColaborador'] = idColaborador != 0 ? idColaborador : null;
		jsonData['dataRequisicao'] = dataRequisicao != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataRequisicao!) : null;
		jsonData['situacao'] = RequisicaoInternaCabecalhoDomain.setSituacao(situacao);
		
		var requisicaoInternaDetalheModelLocalList = []; 
		for (RequisicaoInternaDetalheModel object in requisicaoInternaDetalheModelList ?? []) { 
			requisicaoInternaDetalheModelLocalList.add(object.toJson); 
		}
		jsonData['requisicaoInternaDetalheModelList'] = requisicaoInternaDetalheModelLocalList;
		jsonData['viewPessoaColaboradorModel'] = viewPessoaColaboradorModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idColaborador = plutoRow.cells['idColaborador']?.value;
		dataRequisicao = Util.stringToDate(plutoRow.cells['dataRequisicao']?.value);
		situacao = plutoRow.cells['situacao']?.value != '' ? plutoRow.cells['situacao']?.value : 'Aberta';
		requisicaoInternaDetalheModelList = [];
		viewPessoaColaboradorModel = ViewPessoaColaboradorModel();
		viewPessoaColaboradorModel?.nome = plutoRow.cells['viewPessoaColaboradorModel']?.value;
	}	

	RequisicaoInternaCabecalhoModel clone() {
		return RequisicaoInternaCabecalhoModel(
			id: id,
			idColaborador: idColaborador,
			dataRequisicao: dataRequisicao,
			situacao: situacao,
			requisicaoInternaDetalheModelList: requisicaoInternaDetalheModelListClone(requisicaoInternaDetalheModelList!),
		);			
	}

	requisicaoInternaDetalheModelListClone(List<RequisicaoInternaDetalheModel> requisicaoInternaDetalheModelList) { 
		List<RequisicaoInternaDetalheModel> resultList = [];
		for (var requisicaoInternaDetalheModel in requisicaoInternaDetalheModelList) {
			resultList.add(
				RequisicaoInternaDetalheModel(
					id: requisicaoInternaDetalheModel.id,
					idRequisicaoInternaCabecalho: requisicaoInternaDetalheModel.idRequisicaoInternaCabecalho,
					idProduto: requisicaoInternaDetalheModel.idProduto,
					quantidade: requisicaoInternaDetalheModel.quantidade,
				)
			);
		}
		return resultList;
	}

	
}